pinot-broker/src/main/java/org/apache/pinot/broker/routing/FederatedRoutingManager.java

```diff
@@ -0,0 +1,343 @@
+    Set<String> primaryServingInstances = _primaryRoutingManager.getServingInstances(tableNameWithType);
+    if (primaryServingInstances == null) {
+      primaryServingInstances = Collections.emptySet();
+    }
+    // Combine with serving instances from secondary routing managers
+    Set<String> combinedServingInstances = new HashSet<>(primaryServingInstances);
+    for (BrokerRoutingManager secondaryRoutingManager : _secondaryRoutingManagers) {
+      try {
+        Set<String> secondaryServingInstances = secondaryRoutingManager.getServingInstances(tableNameWithType);
+        if (secondaryServingInstances != null) {
+          combinedServingInstances.addAll(secondaryServingInstances);
+        }
+      } catch (Exception e) {
+        LOGGER.error("Error getting serving instances from secondary routing manager for table {}",
+            tableNameWithType, e);
+      }
+    }
+
+    return combinedServingInstances.isEmpty() ? null : combinedServingInstances;
+  }
+
+  @Override
+  public List<String> getSegments(BrokerRequest brokerRequest) {
+    // Get segments from primary routing manager
+    List<String> primarySegments = _primaryRoutingManager.getSegments(brokerRequest);
+
+    // Combine with segments from secondary routing managers
+    List<String> combinedSegments = new ArrayList<>(primarySegments);
+    for (BrokerRoutingManager secondaryRoutingManager : _secondaryRoutingManagers) {
+      try {
+        List<String> secondarySegments = secondaryRoutingManager.getSegments(brokerRequest);
+        combinedSegments.addAll(secondarySegments);
+      } catch (Exception e) {
+        LOGGER.error("Error getting segments from secondary routing manager", e);
+      }
+    }
+
+    return combinedSegments;
+  }
+
+  @Override
+  public TablePartitionReplicatedServersInfo getTablePartitionReplicatedServersInfo(String tableNameWithType) {
+    // Try to get table partition replicated servers info from primary routing manager first
+    TablePartitionReplicatedServersInfo primaryInfo =
+        _primaryRoutingManager.getTablePartitionReplicatedServersInfo(tableNameWithType);
+    if (primaryInfo != null) {
+      return primaryInfo;
+    }
+
+    // Try to get table partition replicated servers info from secondary routing managers
+    for (BrokerRoutingManager secondaryRoutingManager : _secondaryRoutingManagers) {
+      try {
+        TablePartitionReplicatedServersInfo secondaryInfo =
+            secondaryRoutingManager.getTablePartitionReplicatedServersInfo(tableNameWithType);
+        if (secondaryInfo != null) {
+          return secondaryInfo;
+        }
+      } catch (Exception e) {
+        LOGGER.error(
+            "Error getting table partition replicated servers info from secondary routing manager for table {}",
+            tableNameWithType, e);
+      }
+    }
+
+    return null;
+  }
+
+  public RoutingManager getPrimaryRoutingManager(Map<String, String> queryOptions) {
```

Comment:
this is confusing for me. Here RoutingManager is returning another routing manager, and I think this can be quite error prone long term.

Maybe the right abstraction here is a "RoutingManagerProvider" concept?